src/fs.rs

```diff
@@ -97,7 +112,217 @@ fn write_directory_contents(dir: &Directory, fd: &OwnedFd, repo: &Repository) ->
     Ok(())
 }
 
+// NB: hardlinks not supported
 pub fn write_to_path(repo: &Repository, dir: &Directory, output_dir: &Path) -> Result<()> {
     let fd = openat(CWD, output_dir, OFlags::PATH | OFlags::DIRECTORY, 0.into())?;
     write_directory_contents(dir, &fd, repo)
 }
+
+pub struct FilesystemReader<'repo> {
+    st_dev: u64,
+    repo: Option<&'repo Repository>,
+    inodes: HashMap<u64, Rc<Leaf>>,
+    root_mtime: i64,
+}
+
+impl<'repo> FilesystemReader<'repo> {
+    fn read_xattrs(&mut self, fd: &OwnedFd) -> Result<BTreeMap<Box<OsStr>, Box<[u8]>>> {
+        // flistxattr() and fgetxattr() don't with with O_PATH fds, so go via /proc/self/fd. Note:
+        // we want the symlink-following version of this call, which produces the correct behaviour
+        // even when trying to read xattrs from symlinks themselves.  See
+        // https://gist.github.com/allisonkarlitskaya/7a80f2ebb3314d80f45c653a1ba0e398
+        let filename = proc_self_fd(fd);
+
+        let mut xattrs = BTreeMap::new();
+
+        let names_size = listxattr(&filename, &mut [])?;
+        let mut names = vec![0; names_size];
```

Comment:
Why isn't this `vec![0u8; names_size]` to start?

Reply:
This API is annoying and awful.  We don't know the names size in advance, so we have to query it in order to allocate the buffer.  Fair enough: you do that by submitting a request with an empty buffer and get the correct size returned.  Then you allocate and do it again.  There's a race, of course, if the xattrs changed on disk in the meantime.

I was using a static buffer here at first (which I also shared with the static buffer I use for reading the actual values) but that brings us to the next problem: this API is defined in terms of `c_char` which is `i8` — not `u8`.  A slice of `c_char` might sound like it kinda makes sense when you're dealing with a C API, but when you write it (equivalently) as `[i8]` you notice how bizarre it is.  I can't really see this as anything other than a rustix bug, to be honest...